dali/ft/filecopy.cpp

```diff
@@ -2596,6 +2625,190 @@ void FileSprayer::pushParts()
+    else
+        uri.append(".blob");
+    uri.appendf(".core.windows.net/%s", storagePlane->queryStorageContainer());
+
+    return uri;
+}
+
+using namespace Azure::Storage;
+using namespace Azure::Storage::Files;
+using namespace Azure::Storage::Blobs;
+
+bool FileSprayer::transferUsingAPI()
+{
+    // N.B. ensure api copy is possible by making sure useAPICopy() has returned true
+    // TODO:
+    // 1) bool needCalcCRC = calcCRC(); // How should CRC be handled?
+    // 2) file renames
+    // 3) allow access using "AZURE_ACCOUNT_KEY" env variable
+    // 4) allow access using connection string
+    // 5) support striping with multiple containers/file shares
+    // 6) handle overwrite flag
+    // 7) authenticate using managed identity
+
+    // Source
+    StringBuffer sourceClusterName, sourceBaseURI, sourceSAStoken;
+    distributedSource->getClusterName(0, sourceClusterName);
+    Owned<IStoragePlane> sourcePlane = getDataStoragePlane(sourceClusterName.str(), false);
+    buildAzureStorageURIBase(sourcePlane, sourceBaseURI);
+    sourcePlane->querySASToken(sourceSAStoken);
+    const char * sourcePrefix = sourcePlane->queryPrefix();
+    unsigned sourcePrefixLength = isEmptyString(sourcePrefix)?0:strlen(sourcePrefix);
+
+    // Target
+    StringBuffer targetClusterName, targetBaseURI, targetSAStoken;
+    distributedTarget->getClusterName(0, targetClusterName);
+    Owned<IStoragePlane> targetPlane = getDataStoragePlane(targetClusterName.str(), false);
+    buildAzureStorageURIBase(targetPlane, targetBaseURI);
+    targetPlane->querySASToken(targetSAStoken);
+    const char * targetPrefix = targetPlane->queryPrefix();
+    unsigned targetPrefixLength = isEmptyString(targetPrefix)?0:strlen(targetPrefix);
+    auto targetType = targetPlane->getStorageType();
+
+    std::vector<Shares::StartFileCopyOperation> fileCopyOp;
+    std::vector<StartBlobCopyOperation> blobCopyOp;
+    std::vector<unsigned> copyingInProgress;
+    if (targetType==StorageType::StorageTypeAzureFile)
+        fileCopyOp.reserve(partition.length());
+    else
+        blobCopyOp.reserve(partition.length());
+    // Submit 'copy' operations
+    try
+    {
+         ForEachItemIn(idx, partition)
+         {
+            PartitionPoint & cur = partition.item(idx);
+            OutputProgress & curprogress= progress.item(idx);
+
+            if (curprogress.status==OutputProgress::StatusCopied)
+                continue;
+
+            StringBuffer sourceURI(sourceBaseURI);
+            StringBuffer sourcePath;
+            cur.inputName.getPathInStoragePlane(sourcePath);
+            StringBuffer tmp;
+            sourceURI.appendf("/%s/%s",  encode_url(tmp, sourcePath.str()).str(), sourceSAStoken.str());
+
+            StringBuffer targetURI(targetBaseURI);
+            StringBuffer targetPath;
+            cur.outputName.getPathInStoragePlane(targetPath);
+            targetURI.appendf("/%s/%s", encode_url(tmp.clear(), targetPath.str()).str(), targetSAStoken.str());
```

Comment:
the source and target code is very similar, and looks like it should be commoned up.